### ISSUE TITLE
Linphone and Raspberry Pi

### ISSUE THREAD

**Issue by hsmptg** [BelledonneCommunications/linphone-sdk#80]
Hi
It seems that Simon Morlat modified the page "Linphone and Raspberry Pi" on 2020/09/06: https://wiki.linphone.org/xwiki/wiki/public/view/Linphone/Linphone%20and%20Raspberry%20Pi/
It can be read that: "The procedure was lastly tested on Raspbian 10.4 (August 2020)"
But I just did that in a Raspberry Pi 3 with the last Raspbian and followed all the steps described in this page but I got the following errors: https://pastebin.com/0PQRt5YL

Then I installed the Linphone using "sudo apt install linphone" but it was just version 3.12.0!
With this version I have the problem already described in many places of having a "segmentation fault" if I run a "answer" command in linphonec or linphonecsh.

So, back to square zero, where I have to compile the newest version hoping that this problem is already solved.
Can you help me avoiding the errors I got during compilation (please check the pastebin link)?

I really loved this application (both audio and video worked 100%) but is a pity I can not answer calls using linphonec (for my application that is crucial)! :(
Regards

**Comment by Viish** [comment]
Hi,
Yes currently libvpx doesn't build on raspberry pi. 
Just disable it with `cmake .. -DENABLE_VPX=NO`.
Cheers,

**Comment by hsmptg** [comment]
Hi
Many thanks for your quick suggestion.
As you can check I disabled the libvpx: https://pastebin.com/SAd42mJv

The compilation seems to go a bit further but I get new errors: https://pastebin.com/EKDKh8qy
Can I say that this time the problem is on ffmpeg?
I presume that this lib can not be avoided, right?

In another project I was able to compile it separately with success using this method: https://pimylifeup.com/compiling-ffmpeg-raspberry-pi/
Can I do that and disable it in the linphone compilation?
Regards

**Comment by Viish** [comment]
Yes you can saely disable ffmpeg build using -DENABLE_FFMPEG=NO.

**Comment by hsmptg** [comment]
Hi
Even with this new lib disable I still get errors! :(
If you look to both "pastebins" I shared, the lines that show errors have a common pattern:
/usr/lib/gcc/arm-linux-gnueabihf/8/include/arm_neon.h:XXXX:1: error: inlining failed in call to always_inline ‘vXXXX_f32’: target specific option mismatch
So the problem seems to be related with this "arm_neon thing"!?!
I tried to include a -DNEON_SIMD_ENABLE=OFF but the problem remains.
I saw a post in the Internet suggesting the inclusion of -mfpu=neon, but where I could do that?
Any suggestions?
Regards

**Comment by dwkane** [comment]
I just got my first build to work on Raspberry Pi.  I basically turned everything off except for the following.  I don't need video so all that is disabled also.

ENABLE_ASSETS
ENABLE_MBEDTLS
ENABLE_SPEEX
ENABLE_SQLITE
ENABLE_SRTP
ENABLE_TOOLS

Then, before you run the build, you need to remove a line about dummy.sh in CMakeFiles/sdk.dir/build.make (around line 75)
(this is my build path)

`	cd /home/pi/lin_build/linphone-sdk/build-rasp/desktop && /home/pi/lin_build/linphone-sdk/cmake/dummy.sh`

That's how I got it to work.  Once you do, then you can go back and switch on options one at a time to see which one will break the build.

**Comment by Viish** [comment]
I tried on an raspbian (freshly installed & updated) and the following was enough for me to build the SDK (`feature/python_wrapper` branch but should work as well on `master`) : 
`cmake .. -DENABLE_MKV=NO -DENABLE_OPUS=NO -DENABLE_VPX=NO -DENABLE_WEBRTC_AEC=NO`

**Comment by hsmptg** [comment]
> I tried on an raspbian (freshly installed & updated) and the following was enough for me to build the SDK (`feature/python_wrapper` branch but should work as well on `master`) :
> `cmake .. -DENABLE_MKV=NO -DENABLE_OPUS=NO -DENABLE_VPX=NO -DENABLE_WEBRTC_AEC=NO`

Dear Viish
Thanks you for your reply!
Even with your cmake command I still have errors in the compilation!
Can you please tell me the model of the Raspberry Pi and the exact version of the Raspbian that you used:
- 2020-08-20-raspios-buster-armhf-full.img
- 2020-08-20-raspios-buster-armhf.img
- 2020-08-20-raspios-buster-armhf-lite.img
Regards

**Comment by Viish** [comment]
I'm using a Raspberry Pi 4 (4Go Ram) with a freshly installed Raspbian Lite SD card from RPi Imager and I updated my system upon first boot (apt-get update && apt-get upgrade)

**Comment by hsmptg** [comment]
I was using a Raspberry Pi 3B where I had all the problems I described!
When I changed to a Raspberry Pi 4 (4GB) I was able to compile after some additional minor pre installations (doxygen and pystache).
I am now trying to work with this new linphonec and linphonecsh and check if this time I do not have the segmentation fault after an "answer command" I had in version 3.12.0 (the old one that can be installed through Raspbian repository).

**Comment by hsmptg** [comment]
In my last post I wrote "trying to work with the new linphonec..." because when I run it, the following errors are displayed:
```
pi@rpi-4gb:~ $ ~/linphone-sdk/build-raspberry/linphone-sdk/desktop/bin/linphonec
2020-11-20 18:21:28:688 bctbx-error-bctbx_file_open: Error open No such file or directory
2020-11-20 18:21:28:689 liblinphone-error-Unable to open linphone database with uri "/home/pi/.local/share/linphone/linphone.db" and backend Sqlite3
Warning: video is disabled in linphonec, use -V or -C or -D to enable.
linphonec>
```
Notice that unlike the [page](https://wiki.linphone.org/xwiki/wiki/public/view/Linphone/Linphone%20and%20Raspberry%20Pi/) says, the binary is present at ~/linphone-sdk/build-raspberry/linphone-sdk/desktop/bin and not at OUTPUT/no-ui/bin.

But, back to the error in linphonec, what is that /home/pi/.local/share/linphone/linphone.db, that in fact in not present in my system?

**Comment by Viish** [comment]
Don't worry, it's just pre-defined path for database storage of chat messages.
Either create the directory (`/home/pi/.local/share/linphone/`), and the file will be created, or configure another path in your config file with:
```
[storage]
uri=<path>
```

**Comment by hsmptg** [comment]
Dear Viish
I think I am almost there! :)
Your tip removed the error and I could run linphonec without any message:

```
pi@rpi-4gb:~/linphone-sdk/build-raspberry/linphone-sdk/desktop/bin $ ./linphonec -V
linphonec>
```

Then I added a proxy to my Asterisk server:

```linphonec> proxy list
No proxies defined
linphonec> proxy add
Adding new proxy setup. Hit ^D to abort.
Enter proxy sip address: sip:10.0.5.25
Your identity for this proxy: sip:3000@10.0.5.25
Do you want to register on this proxy (yes/no): yes
Specify register expiration time in seconds (default is 600):
Expiration: 600 seconds
Specify route if needed:
No route specified.
--------------------------------------------
sip address: sip:10.0.5.25
route:
identity: sip:3000@10.0.5.25
register: yes
expires: 600
registered: no
--------------------------------------------
Accept the above proxy configuration (yes/no) ?: yes
Proxy added.
linphonec> 2020-11-21 11:30:45:128 liblinphone-error-Given algorithm is not correct. Set algorithm failed
2020-11-21 11:30:45:128 liblinphone-error-Given algorithm is not correct. Add algorithm failed

Password for 3000 on mypbx.local: 1234
2020-11-21 11:31:05:648 liblinphone-error-Given algorithm is not correct. Set algorithm failed
2020-11-21 11:31:05:648 liblinphone-error-Given algorithm is not correct. Add algorithm failed

linphonec>
```

Even with this error I was able to send and receive audio calls in the Raspberry Pi, and this time without any segmentation faults! GREAT!!!

Nevertheless the video of my Raspberry Pi camera that I was able to see in version 3.12.0 (the old one that can be installed through Raspbian repository) this time do not appear! :/
Do you think this is related with the "algorithm is not correct. Set algorithm failed" error or is due the compilation (remember that the cmake I used did not included the -DENABLE_OPENH264=ON and -DENABLE_FFMPEG=ON)?

Regards

**Comment by Viish** [comment]
No `"algorithm is not correct. Set algorithm failed"` is a "new" feature from the SDK that wasn't added in linphonec, so that's why.
And anyway, it won't impact video, it's merely a proxy authentication feature.
Regarding the camera, I don't know. Can you start linphonec in verbose mode (`./linphonec -d 6`) and attach full logs please?

**Comment by hsmptg** [comment]
Hi
Here is the log of a PuTTY ssh session where after I run linphonec in my Raspberry Pi (extension 300), I answered a call from my PC (extension 4000): https://pastebin.com/1ZsuZMfQ
Once again the audio worked 100% but not the video (the Raspberry Pi cam I mean).
Do you see anything suspicious?
Since in the log I saw several messages saying "Could not find encoder for <codec>", I tried to install H264 with:
```
git clone https://github.com/cisco/openh264.git \
  && cd openh264 \
  && make \
  && sudo make install
```
But the problem remained as well with message "Could not find encoder for H264".
Regards

**Comment by Viish** [comment]
Indeed it doesn't detect the camera device.
Do you have installed bcm2835-v4l2 and make sure it was loaded with modprobe?

**Comment by hsmptg** [comment]
Yes, I even test it by tacking a picture:
```
sudo modprobe bcm2835-v4l2
v4l2-ctl --set-fmt-video=width=640,height=480,pixelformat=3
v4l2-ctl --stream-mmap=3 --stream-count=1 --stream-to=test1.jpg
```
Inside linphonec, I see a webcam. Well it seems to be:
```
pi@rpi-4gb:~/linphone-sdk/build-raspberry/linphone-sdk/desktop/bin $ ./linphonec -V
2020-11-21 14:15:28:963 liblinphone-error-Given algorithm is not correct. Set algorithm failed
2020-11-21 14:15:28:963 liblinphone-error-Given algorithm is not correct. Add algorithm failed
2020-11-21 14:15:28:963 liblinphone-error-Given algorithm is not correct. Set algorithm failed
2020-11-21 14:15:28:963 liblinphone-error-Given algorithm is not correct. Add algorithm failed
linphonec> webcam list
0: StaticImage: Static picture
linphonec>
```
The [cmake I used](https://github.com/BelledonneCommunications/linphone-sdk/issues/80#issuecomment-729757492) would allow working the Raspberry Pi camera?
Regards

**Comment by Viish** [comment]
Do you have disabled video with -DENABLE_VIDEO=NO?

**Comment by hsmptg** [comment]
As I said I used a cmake exactly as you wrote in a previous post:
```
cmake .. -DENABLE_MKV=NO -DENABLE_OPUS=NO -DENABLE_VPX=NO -DENABLE_WEBRTC_AEC=NO
```
Any of these "flags" disable the video?

**Comment by Viish** [comment]
Ah OK problem is you don't have any video encoder/decoder available, hence no video.
So your ffmpeg build doesn't work with our SDK for some reason.

**Comment by hsmptg** [comment]
Hi
I'm back to report my last steps!

When you wrote "your ffmpeg build" I realised that I did not yet had installed the ffmpeg.
I did that but the "no camera" issue still occurred.

Then, I thought that maybe the ffmpeg installation must be done before the linphone compilation.
So I repeat those two steps in that order but the "no camera" issue still occurred (here is the log: https://pastebin.com/ANLB4yTM).

Finally I installed the old version of linphone (version: 3.12.0) using "sudo apt install linphone" and after a quick test the camera worked as expected!

Any suggestion on what to do now?
It is a pitty that this 3.12.0 have that segmentation fault after the "answer command" otherwise was enough for me!

**Comment by hsmptg** [comment]
By the way, [in this page](https://wiki.linphone.org/xwiki/wiki/public/view/Linphone/Linphone%20and%20Raspberry%20Pi/) the cmake command have all those "-DENABLExxx" equal to ON or OFF.
In the cmake you suggested me [here](https://github.com/BelledonneCommunications/linphone-sdk/issues/80#issuecomment-729757492), all the "-DENABLExxx" are equal to NO.
Is this a typo (and instead of NO should be ON) or is correct (and NO is equal to OFF)?

**Comment by Viish** [comment]
> Is this a typo (and instead of NO should be ON) or is correct (and NO is equal to OFF)?

No it's not a typo, OFF = NO.

**Comment by Viish** [comment]
Maybe you build ffmpeg ithout h264 encoder/decoder?

**Comment by hsmptg** [comment]
> Maybe you build ffmpeg ithout h264 encoder/decoder?

I installed h264 using this:
```
git clone https://github.com/cisco/openh264.git
cd openh264
make
sudo make install
```
I then installed ffmpeg using this:
```
git clone --depth 1 https://github.com/FFmpeg/FFmpeg.git ~/FFmpeg \
  && cd ~/FFmpeg \
  && ./configure \
    --extra-cflags="-I/usr/local/include" \
    --extra-ldflags="-L/usr/local/lib" \
    --extra-libs="-lpthread -lm -latomic" \
    --arch=armel \
    --enable-gmp \
    --enable-gpl \
    --enable-libaom \
    --enable-libass \
    --enable-libdav1d \
    --enable-libdrm \
    --enable-libfdk-aac \
    --enable-libfreetype \
    --enable-libkvazaar \
    --enable-libmp3lame \
    --enable-libopencore-amrnb \
    --enable-libopencore-amrwb \
    --enable-libopus \
    --enable-librtmp \
    --enable-libsnappy \
    --enable-libsoxr \
    --enable-libssh \
    --enable-libvorbis \
    --enable-libvpx \
    --enable-libzimg \
    --enable-libwebp \
    --enable-libx264 \
    --enable-libx265 \
    --enable-libxml2 \
    --enable-mmal \
    --enable-nonfree \
    --enable-omx \
    --enable-omx-rpi \
    --enable-version3 \
    --target-os=linux \
    --enable-pthreads \
    --enable-openssl \
    --enable-hardcoded-tables \
  && make -j$(nproc) \
  && sudo make install
```
In fact I did all the procedure described in this [page](https://pimylifeup.com/compiling-ffmpeg-raspberry-pi/). 

Please clarify me this: The need of the h264 and ffmpeg is during the compilation AND execution of linphonec, or just during execution?

Because if it is just during execution, the proof that all that is installed, is that linphonec v3.12.0 works (besides the segmentation fault during the answer command).

**Comment by hsmptg** [comment]
I just noticed that if I run ffmpeg from whatever directory, it is recognised!
If I do the same for h264enc or h264dec, the command is not found. It seems that despite I did a "sudo make install" at the end of the openh264 instalation, the binaries are only present in /home/pi/openh264
Do you think this can be a problem?

**Comment by hsmptg** [comment]
One more thing...
In the compiled linphonec I got this:
```
linphonec> webcam list
0: StaticImage: Static picture
```
But in version 3.12.0 I got this:
```
linphonec> webcam list
0: V4L2: /dev/video0
1: StaticImage: Static picture
```
So the problem is, why the compiled version do not recognise the Raspberry camera?
Do you think is a codec issue (h264, ffmpeg) or is a device issue?

By the way, I installed h264 using this [other method](http://jollejolles.com/installing-ffmpeg-with-h264-support-on-raspberry-pi/) but the problem remains! :(

**Comment by hsmptg** [comment]
The problem of /dev/video0 not being shown seems not to be new as can be see [here](https://unix.stackexchange.com/questions/576761/linphone-does-not-detect-v4l2loopback-devices) and [here](https://unix.stackexchange.com/questions/450951/v4l2loopback-devices-not-showing-up-in-linphone-webcam-list).

Yourself warned for that problem in your [webpage](https://wiki.linphone.org/xwiki/wiki/public/view/Linphone/Linphone%20and%20Raspberry%20Pi/).
But I did what you suggested and nevertheless the problem remains!

Any additional suggestion on what to do in order to get /dev/video0 listed as a webcam inside linphonec?
Regards

**Comment by Viish** [comment]
Try the following:

- Edit the cmake-builder/builders/ffmpeg.cmake and add the following after `"--enable-encoder=mjpeg":`
```
"--enable-gpl"
"--enable-nonfree"
"--enable-libvpx"
"--enable-neon"
```
- Run the cmake command again with: `cmake .. -DENABLE_FFMPEG=ON`
- Build again: `make -j2`

**Comment by hsmptg** [comment]
I did all you said and could compile without errors!
But once again I get this:
```
pi@rpi-4gb:~/linphone-sdk/build-raspberry/linphone-sdk/desktop/bin $ ./linphonec -V
2020-11-22 19:59:53:151 liblinphone-error-Given algorithm is not correct. Set algorithm failed
2020-11-22 19:59:53:151 liblinphone-error-Given algorithm is not correct. Add algorithm failed
2020-11-22 19:59:53:151 liblinphone-error-Given algorithm is not correct. Set algorithm failed
2020-11-22 19:59:53:151 liblinphone-error-Given algorithm is not correct. Add algorithm failed
linphonec> webcam list
0: StaticImage: Static picture
```
The log in linphonec till executing the "webcam list" command, can be seen [here](https://pastebin.com/atLjtr4U).

**Comment by hsmptg** [comment]
Dear Viish
Let me correct something. (Sorry)
When I said in my last post that I followed all your commands, I realised meanwhile that was not entirely true, since instead of:
```
cmake .. -DENABLE_FFMPEG=ON
```
What I did was to add this new flag to your previous cmake command, that is:
```
cmake .. -DENABLE_MKV=NO -DENABLE_OPUS=NO -DENABLE_VPX=NO -DENABLE_WEBRTC_AEC=NO -DENABLE_FFMPEG=ON
```
Meanwhile I preform a new compilation with just ```cmake .. -DENABLE_FFMPEG=ON``` but this time I got the same kind of compilation errors I got some days ago (those involving arm_neon.h):
```
In file included from /home/pi/linphone-sdk/external/opus/celt/arm/celt_neon_intr.c:37:
/usr/lib/gcc/arm-linux-gnueabihf/8/include/arm_neon.h:7997:1: error: inlining failed in call to always_inline ‘vmlaq_lane_f32’: target specific option mismatch
 vmlaq_lane_f32 (float32x4_t __a, float32x4_t __b, float32x2_t __c, const int __d)
 ^~~~~~~~~~~~~~
/home/pi/linphone-sdk/external/opus/celt/arm/celt_neon_intr.c:189:11: note: called from here
    SUMM = vmlaq_lane_f32(SUMM, YY[0], XX_2, 0);
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by Viish** [comment]
Actually you got it right the first time.
As you already did `cmake .. -DENABLE_MKV=NO -DENABLE_OPUS=NO -DENABLE_VPX=NO -DENABLE_WEBRTC_AEC=NO` previously, doing again `cmake .. -DENABLE_FFMPEG=ON` adds the change to the previous configuration, so it is as if you did `cmake .. -DENABLE_MKV=NO -DENABLE_OPUS=NO -DENABLE_VPX=NO -DENABLE_WEBRTC_AEC=NO -DENABLE_FFMPEG=ON` which was the objective.
Obviously it won't miraculously build now opus etc...

Can you try the following? `cmake .. -DENABLE_MKV=NO -DENABLE_OPUS=NO -DENABLE_VPX=NO -DENABLE_WEBRTC_AEC=NO -DENABLE_FFMPEG=ON -DENABLE_V4L=ON`

**Comment by hsmptg** [comment]
Since I already changed so many things, can you tell me the complete cmake command to use in a fresh installation? Can I use this last one?
Regards

**Comment by Viish** [comment]
> Can I use this last one?

Yes

**Comment by hsmptg** [comment]
Hi
We are making progresses because now the "webcam list" shows the "V4L2: /dev/video0"!!! :D
```
pi@rpi-4gb:~/linphone-sdk/build-raspberry/linphone-sdk/desktop/bin $ ./linphonec -V
2020-11-23 10:40:13:219 liblinphone-error-Given algorithm is not correct. Set algorithm failed
2020-11-23 10:40:13:219 liblinphone-error-Given algorithm is not correct. Add algorithm failed
2020-11-23 10:40:13:219 liblinphone-error-Given algorithm is not correct. Set algorithm failed
2020-11-23 10:40:13:219 liblinphone-error-Given algorithm is not correct. Add algorithm failed
linphonec> 2020-11-23 10:40:13:235 mediastreamer-error-Could not open display :0
webcam list
0: V4L2: /dev/video0
1: StaticImage: Static picture
linphonec>
```
But if I make a “call 4000@10.0.5.25” (a call to a linphone I have in my PC) the video of the camera do not show up. If I do exactly the same in version 3.12.0 everything works fine. Here are those two tries. First the working version 3.12.0:
```
pi@rpi-4gb:~/linphone-sdk/build-raspberry/linphone-sdk/desktop/bin $ linphonec -V
2020-11-23 10:18:41:343 ortp-error-Connection to the pulseaudio server failed
Ready
linphonec> Refreshing on sip:3000@10.0.5.25...
linphonec> 2020-11-23 10:18:41:592 ortp-error-Could not open display :0
Registration on sip:10.0.5.25 successful.
linphonec> call 4000@10.0.5.25
Establishing call id to sip:4000@10.0.5.25, assigned id 1
Contacting sip:4000@10.0.5.25
linphonec> Call 1 to sip:4000@10.0.5.25 in progress.
linphonec> Remote ringing.
linphonec> Remote ringing...
linphonec> Call 1 to sip:4000@10.0.5.25 ringing.
Call 1 with sip:4000@10.0.5.25 connected.
Call answered by sip:4000@10.0.5.25
linphonec> 2020-11-23 10:19:00:753 ortp-error-no such method on filter MSVp8Dec, fid=16389 method index=7
2020-11-23 10:19:00:754 ortp-error-Could not open display :0
Media streams established with sip:4000@10.0.5.25 for call 1 (video).
Call terminated.
linphonec> Call 1 with sip:4000@10.0.5.25 ended (Unknown error).
2020-11-23 10:19:07:341 ortp-error-Could not open display :0
quit
Terminating...
Unregistration on sip:10.0.5.25 done.
linphonec>
```
Now the compiled version where I don't get any video:
```
pi@rpi-4gb:~/linphone-sdk/build-raspberry/linphone-sdk/desktop/bin $ ./linphonec -V
2020-11-23 10:20:54:209 liblinphone-error-Given algorithm is not correct. Set algorithm failed
2020-11-23 10:20:54:209 liblinphone-error-Given algorithm is not correct. Add algorithm failed
2020-11-23 10:20:54:209 liblinphone-error-Given algorithm is not correct. Set algorithm failed
2020-11-23 10:20:54:209 liblinphone-error-Given algorithm is not correct. Add algorithm failed
linphonec> 2020-11-23 10:20:54:225 mediastreamer-error-Could not open display :0
call 4000@10.0.5.25
Establishing call id to sip:4000@10.0.5.25, assigned id 1
Call 1 to sip:4000@10.0.5.25 in progress.
linphonec> Call 1 to sip:4000@10.0.5.25 ringing.
Call 1 with sip:4000@10.0.5.25 connected.
Media streams established with sip:4000@10.0.5.25 for call 1 (audio).
2020-11-23 10:21:24:168 liblinphone-error-Could not resolv /home/pi/.linphone.ecstate: No such file or directory
Call 1 with sip:4000@10.0.5.25 ended (Unknown error).
2020-11-23 10:21:24:231 mediastreamer-error-Could not open display :0
quit
Terminating...
```
Note that both say "Could not open display :0", but while version 3.12.0 shows "Media streams established with sip:4000@10.0.5.25 for call 1 (video)", the compiled version only shows "Media streams established with sip:4000@10.0.5.25 for call 1 (audio)".

Here you can see the full logs (using -d 6) for [version 3.12.0](https://pastebin.com/mMztbmNL) and [compiled version](https://pastebin.com/LyWr3LUj).
Regards

**Comment by Viish** [comment]
Indeed, you still don't have any video encoder/decoder so video is still disabled.
It seems we don't support VP8 codec through FFMPEG, so you can revert the changes in `cmake-builder/builders/ffmpeg.cmake`.
I'll see what I can do.

**Comment by hsmptg** [comment]
Many thanks for all your effort!
I really appreciate all the work you have done trying to solve this!

Let me tell you the two projects where I intended to use linphonec in my home automation system:

The first is to convert an old rotary phone into a SIP phone. Since I only need audio here I will use this last compiled version.

The second one is a door videophone. Here I need video but for that I can use version 3.12.0 that has the segmentation fault on answer, something that I will not use since I do not need to call the videophone (just the other way around). Even in the case I need to call the videophone, I could send some kind of message to the Raspberry to initiate the video call from there.

The problem now is that I would like to use linphonec in both cases not on a Raspberry Pi 4 but on a Raspberry Pi Zero W and Raspberry Pi 3B respectively. But on a Raspberry Pi Zero W the compilation takes ages and I never could finish it without errors (maybe due to a lack of memory issue). So my question is:

Do you know how can I transfer now to it the compiled version I got in the Raspberry Pi 4? I tried to copy the linphonec binary and got errors concerning the libraries. How can I transfer all that is needed to have a working linphonec/linphonecsh?

Best regards

**Comment by Viish** [comment]
Actually I don't know why I disabled VPX build, I just tried again and it went fine.
So you can try with: `cmake .. -DENABLE_MKV=NO -DENABLE_OPUS=NO -DENABLE_VPX=ON -DENABLE_WEBRTC_AEC=NO -DENABLE_FFMPEG=ON`

Regarding the pi zero, I think the best way would be to build the python wrapper (switch linphone-sdk to `feature/python_wrapper` branch, update submodules and add `-DENABLE_PYTHON_WRAPPER=ON` to the cmake options).
Once built, you will have a pylinphone-<something>.whl file created in `<build-dir>/linphone-sdk/desktop/`.
Send that file your your Pi ZW and install it with: `pip3 install pylinphone*.whl`

It won't package linphonec/linphonecsh tools but you'll be able to create your own python script to make what you want.
You have an example there: https://wiki.linphone.org/xwiki/wiki/public/view/Lib/Linphone%20Python%20Wrapper/#HGettingstarted

**Comment by hsmptg** [comment]
Just to tell you that when I tried this last cmake command you wrote ```cmake .. -DENABLE_MKV=NO -DENABLE_OPUS=NO -DENABLE_VPX=ON -DENABLE_WEBRTC_AEC=NO -DENABLE_FFMPEG=ON``` the ```webcam list``` did not included once again the ```V4L2: /dev/video0```! :(

Then i realised that this command did not included the ```-DENABLE_V4L=ON``` that previously solved that problem so I include it and for the 100th time made the compilation, and this time WORKED!!! Now I got the ```V4L2: /dev/video0``` and I have video in my calls!

The proper cmake command (even with no patch in the ```cmake-builder/builders/ffmpeg.cmake```) is therefore:
```
cmake .. -DENABLE_MKV=NO -DENABLE_OPUS=NO -DENABLE_VPX=ON -DENABLE_WEBRTC_AEC=NO -DENABLE_FFMPEG=ON -DENABLE_V4L=ON
```
So I think now I can say: PROBLEM SOLVED!!!
Thanks for all your help!

Time now to start the "other War" you got me into when suggested the use of the Python wrapper! :D
Do you mind if I use this same thread to clarify some future doubts about that?

**Comment by Viish** [comment]
> and this time WORKED

Great! :)

> Do you mind if I use this same thread to clarify some future doubts about that?

No please open a new ticket if needed, this one starts to be quite lengthy ;)

**Comment by Expander1** [comment]
Hi Viish, hsmptg,

I followed this guide and tryed to compile the linphone-sdk. but at the end I ever have some failures... This is the first time to compile a repo for me and I do not know, how it really works... Sorry, maybe I'm wrong at this, but after finish the compilation I cant find in the specific folder the linphonec /csh files...

My System is a Pi 2B+ with boot SD Card and system on USB stick.

My question is, what do I wrong at compilation? Where can I analyse the compilation. Is there a Log file? During the comp I saw some warnings what shows that there are some statement failures???

I'm not new to linux, but not that deep inside that I have knowlage of source compilation...

Maybe some of you can help me to understand...

Thanks

Sven
